### PR TITLE
[TypeScript] Make types more strict in ra-ui-materialui

### PR DESCRIPTION
**Pull request by fzaninotto**
Decrease TS compilation errors in strictNullChecks mode from 216 to 164.  Small steps.

### CODE REVIEW

File: packages/ra-ui-materialui/src/list/datagrid/Datagrid.tsx [hunk 1]
```diff
@@ -259,7 +260,6 @@ export const Datagrid: FC<DatagridProps> = React.forwardRef((props, ref) => {
                                     hasBulkActions,
                                     isRowSelectable,
                                     onSelect,
-                                    resource,
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/list/datagrid/Datagrid.tsx:262]: We no longer need to pass down resource as there is now always a ResourceContext, even in controlled mode.

**Comment by fzaninotto** [comment]
Switching to RFR

File: packages/ra-ui-materialui/src/list/datagrid/ExpandAllButton.tsx [hunk 2]
```diff
@@ -2,19 +2,20 @@ import * as React from 'react';
 import { memo } from 'react';
 import IconButton from '@mui/material/IconButton';
 import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
-import { useTranslate, useExpandAll } from 'ra-core';
+import { useTranslate, useExpandAll, useResourceContext } from 'ra-core';
 import { DatagridClasses } from './useDatagridStyles';
 import clsx from 'clsx';
 
 interface ExpandAllButtonProps {
-    resource: string;
```
Comment (slax57) [inline packages/ra-ui-materialui/src/list/datagrid/ExpandAllButton.tsx:10]: same

File: packages/ra-ui-materialui/src/list/datagrid/DatagridHeader.tsx [hunk 3]
```diff
@@ -189,7 +186,6 @@ export interface DatagridHeaderProps<RecordType extends RaRecord = any> {
     data?: RecordType[];
     onSelect?: (ids: Identifier[]) => void;
     onToggleItem?: (id: Identifier) => void;
-    resource?: string;
```
[the remark this answers is not in the document]
Comment (fzaninotto) [reply packages/ra-ui-materialui/src/list/datagrid/DatagridHeader.tsx:192]: i don't see in which case users would inject the `resource` themselves, outside of the context of the Datagrid itself (which used to override the resource anyway). So I'm -1 for adding it to the Upgrade guide.

File: packages/ra-ui-materialui/src/list/datagrid/ExpandAllButton.tsx (hunk 2 above)
Comment (slax57) [reply packages/ra-ui-materialui/src/list/datagrid/ExpandAllButton.tsx:10]: Actually this is an internal component only used by our own DatagridHeader, so this doesn't count as a BC

**Review by slax57 — APPROVED** [review]

**Pull request merged**